src/category_theory/subobject.lean

```diff
@@ -0,0 +1,1112 @@
+with morphisms becoming inequalities, and isomorphisms becoming equations.
+-/
+
+/--
+The category of subobjects of `X : C`, defined as isomorphism classes of monomorphisms into `X`.
+-/
+@[derive [partial_order, category]]
+def subobject (X : C) := thin_skeleton (mono_over X)
+
+namespace subobject
+
+/-- Convenience constructor for a subobject. -/
+abbreviation mk {X A : C} (f : A ⟶ X) [mono f] : subobject X :=
+(to_thin_skeleton _).obj (mono_over.mk' f)
+
+/--
+Use choice to pick a representative `mono_over X` for each `subobject X`.
+-/
+noncomputable
+def representative {X : C} : subobject X ⥤ mono_over X :=
+thin_skeleton.from_thin_skeleton _
+
+
+/--
+Starting with `A : mono_over X`, we can take its equivalence class in `subobject X`
+then pick an arbitrary representative using `representative.obj`.
+This is isomorphic (in `mono_over X`) to the original `A`.
+-/
+noncomputable
+def representative_iso {X : C} (A : mono_over X) :
+  representative.obj ((to_thin_skeleton _).obj A) ≅ A :=
+(thin_skeleton.from_thin_skeleton _).as_equivalence.counit_iso.app A
+
+/--
+Use choice to pick a representative underlying object in `C` for any `subobject X`.
+
+Prefer to use the coercion `P : C` rather than explicitly writing `underlying.obj P`.
+-/
+noncomputable
+def underlying {X : C} : subobject X ⥤ C :=
+representative ⋙ mono_over.forget _ ⋙ over.forget _
+
+instance : has_coe (subobject X) C :=
+{ coe := λ Y, underlying.obj Y, }
+
+@[simp] lemma underlying_as_coe {X : C} (P : subobject X) : underlying.obj P = P := rfl
+
+/--
+If we construct a `subobject Y` from an explicit `f : X ⟶ Y` with `[mono f]`,
+then pick an arbitrary choice of underlying object `(subobject.mk f : C)` back in `C`,
+it is isomorphic (in `C`) to the original `X`.
+-/
+noncomputable
+def underlying_iso {X Y : C} (f : X ⟶ Y) [mono f] : (subobject.mk f : C) ≅ X :=
+(mono_over.forget _ ⋙ over.forget _).map_iso (representative_iso (mono_over.mk' f))
+
+/--
+The morphism in `C` from the arbitrarily chosen underlying object to the ambient object.
+-/
+noncomputable
+def arrow {X : C} (Y : subobject X) : (Y : C) ⟶ X :=
+(representative.obj Y).val.hom
+
+@[simp]
+lemma underlying_arrow {X : C} {Y Z : subobject X} (f : Y ⟶ Z) :
+  underlying.map f ≫ arrow Z = arrow Y :=
+over.w (representative.map f)
+
+end subobject
+
+namespace limits
+
+section equalizer
+variables (f g : X ⟶ Y) [has_equalizer f g]
+
+/-- The equalizer of morphisms `f g : X ⟶ Y` as a `subobject X`. -/
+def equalizer_subobject : subobject X :=
+subobject.mk (equalizer.ι f g)
+
+/-- The underlying object of `equalizer_subobject f g` is (up to isomorphism!)
+the same as the chosen object `equalizer f g`. -/
+def equalizer_subobject_iso : (equalizer_subobject f g : C) ≅ equalizer f g :=
+subobject.underlying_iso (equalizer.ι f g)
+
+lemma equalizer_subobject_arrow :
+  (equalizer_subobject f g).arrow = (equalizer_subobject_iso f g).hom ≫ equalizer.ι f g :=
```

Comment:
Do you think this could be more helpful the other way round? And possibly as a simp lemma too?

Reply:
It makes proving
```
lemma equalizer_subobject_arrow_comp :
  (equalizer_subobject f g).arrow ≫ f = (equalizer_subobject f g).arrow ≫ g :=
by simp [equalizer_subobject_arrow, equalizer.condition]
```
below more painful.

I'm a bit uncertain about the complexity function here. The RHS looks more complicated (involving a composition of two morphisms), but on the other hand it is "simpler", being about the "more fundamental" `equalizer.ι` rather than `equalizer_subobject f g).arrow`. This makes me not want a simp lemma either way, and it is more convenient for use immediately below in this direction.

The next lemma
```
lemma equalizer_subobject_arrow' :
  (equalizer_subobject_iso f g).inv ≫ (equalizer_subobject f g).arrow = equalizer.ι f g  
```
is unambiguously simpler on the LHS.

Reply:
Fair enough! I agree the lemma `equalizer_subobject_arrow'` is perfectly fine as it is - let's stick with this one too then.